---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [1]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle


---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [2]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [3]:
from collections import Counter
def degree_distribution(g):
    c = Counter(g.degree().values())
    d = {k:v/nx.number_of_nodes(g) for k, v in c.items()}
    return d

In [4]:
info = dict()
for i in range(5):
    g = P1_Graphs[i]
    temp = [len(degree_distribution(g)), 
            nx.average_clustering(g),
            nx.average_shortest_path_length(g)]
    info[i] = temp
info

{0: [29, 0.03167539146454044, 4.099161161161161],
 1: [7, 0.5642419635919628, 5.089871871871872],
 2: [5, 0.4018222222222227, 9.378702269692925],
 3: [37, 0.03780379975223251, 3.1048046283934134],
 4: [9, 0.0033037037037037037, 5.0785509568313305]}

In [5]:
def graph_identification():
    return ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [95]:
G = nx.read_gpickle('email_prediction.txt')
G.nodes(data=True)

[(0, {'Department': 1, 'ManagementSalary': 0.0}),
 (1, {'Department': 1, 'ManagementSalary': nan}),
 (2, {'Department': 21, 'ManagementSalary': nan}),
 (3, {'Department': 21, 'ManagementSalary': 1.0}),
 (4, {'Department': 21, 'ManagementSalary': 1.0}),
 (5, {'Department': 25, 'ManagementSalary': nan}),
 (6, {'Department': 25, 'ManagementSalary': 1.0}),
 (7, {'Department': 14, 'ManagementSalary': 0.0}),
 (8, {'Department': 14, 'ManagementSalary': nan}),
 (9, {'Department': 14, 'ManagementSalary': 0.0}),
 (10, {'Department': 9, 'ManagementSalary': 0.0}),
 (11, {'Department': 14, 'ManagementSalary': 0.0}),
 (12, {'Department': 14, 'ManagementSalary': 1.0}),
 (13, {'Department': 26, 'ManagementSalary': 1.0}),
 (14, {'Department': 4, 'ManagementSalary': nan}),
 (15, {'Department': 17, 'ManagementSalary': 0.0}),
 (16, {'Department': 34, 'ManagementSalary': 0.0}),
 (17, {'Department': 1, 'ManagementSalary': 0.0}),
 (18, {'Department': 1, 'ManagementSalary': nan}),
 (19, {'Department': 14, 'Ma

In [76]:
dep = [i[1]['Department'] for i in G.nodes(data=True)]
sal = [i[1]['ManagementSalary'] for i in G.nodes(data=True)]
df = pd.DataFrame({'dep': [i[1]['Department'] for i in G.nodes(data=True)], 
                   'sal': [i[1]['ManagementSalary'] for i in G.nodes(data=True)], 
                   'clu': pd.Series(nx.clustering(G)), 
                   'deg': pd.Series(G.degree()), 
                   'clo': pd.Series(nx.closeness_centrality(G, normalized=True)), 
                   'bet': pd.Series(nx.betweenness_centrality(G, normalized=True)), 
                   'pgr': pd.Series(nx.pagerank(G))})
df.head(40)

dep  sal       clu  deg       clo           bet       pgr
0       1  0.0  0.276423   44  0.421991  1.123719e-03  0.001224
1       1  NaN  0.265306   52  0.422360  1.195088e-03  0.001426
2      21  NaN  0.297803   95  0.461490  6.569720e-03  0.002605
3      21  1.0  0.384910   71  0.441663  1.653572e-03  0.001833
4      21  1.0  0.318691   96  0.462152  5.547474e-03  0.002526
5      25  NaN  0.107002  171  0.501484  3.099469e-02  0.004914
6      25  1.0  0.155183  115  0.475805  1.238664e-02  0.003146
7      14  0.0  0.287785   72  0.420156  2.817676e-03  0.002002
8      14  NaN  0.447059   37  0.413151  5.568287e-04  0.001059
9      14  0.0  0.425320   40  0.356196  2.803031e-04  0.001203
10      9  0.0  0.315449   47  0.442472  1.084132e-03  0.001303
11     14  0.0  0.229174   76  0.431218  3.946129e-03  0.002064
12     14  1.0  0.302597   58  0.433151  4.715059e-03  0.001715
13     26  1.0  0.110900  180  0.503050  2.356499e-02  0.004590
14      4  NaN  0.215784   80  0.442068  3.725905e-03  0.002166
15     17  0.0  0.352653   52  0.429875  2.216578e-03  0.001457
16     34  0.0  0.284709   74  0.457990  1.490193e-03  0.001870
17      1  0.0  0.285933  111  0.469791  5.220689e-03  0.002768
18      1  NaN  0.301188   56  0.442068  3.075746e-03  0.001531
19     14  0.0  0.283230   72  0.441462  3.632578e-03  0.001891
20      9  0.0  0.284820   74  0.451992  3.890008e-03  0.002032
21      9  1.0  0.248039  138  0.487814  1.159672e-02  0.003663
22      9  0.0  0.384615   15  0.379262  5.491745e-05  0.000499
23     11  0.0  0.342746   81  0.451359  3.255794e-03  0.002036
24     11  0.0  0.385491   40  0.398170  1.529863e-03  0.001212
25     11  0.0  0.482955   33  0.413858  2.177656e-04  0.000904
26     11  0.0  0.520833   35  0.381508  2.448031e-04  0.000949
27     11  NaN  0.368852   63  0.430258  3.116899e-03  0.001713
28     11  1.0  0.344755  105  0.467292  5.120838e-03  0.002588
29     11  1.0  0.393506   58  0.441462  1.418737e-03  0.001492
...   ...  ...       ...  ...       ...           ...       ...
975     9  0.0  0.000000    1  0.305327  0.000000e+00  0.000177
976    10  0.0  0.766667   27  0.361797  4.074871e-05  0.000769
977    19  0.0  1.000000    4  0.313957  0.000000e+00  0.000282
978    13  0.0  1.000000    5  0.331626  0.000000e+00  0.000272
979    23  0.0  0.200000    7  0.343533  1.192065e-05  0.000330
980    12  0.0  0.400000   12  0.372393  2.143291e-04  0.000467
981    10  0.0  0.766667   27  0.361797  4.074871e-05  0.000769
982    21  0.0  0.000000    1  0.310926  0.000000e+00  0.000173
983    10  0.0  1.000000    3  0.341229  0.000000e+00  0.000214
984    35  NaN  0.333333    3  0.332768  8.620144e-07  0.000220
985     7  0.0  0.000000    1  0.293726  0.000000e+00  0.000177
986    22  0.0  0.666667    3  0.325593  1.541983e-06  0.000225
987    22  NaN  0.803030   12  0.355802  5.920895e-06  0.000428
988    22  0.0  1.000000    2  0.320093  0.000000e+00  0.000201
989     8  NaN  1.000000    4  0.317672  0.000000e+00  0.000276
990    21  0.0  0.621053   20  0.372968  1.279016e-04  0.000641
991    32  NaN  1.000000    4  0.322013  0.000000e+00  0.000244
992     4  NaN  0.000000    3  0.332998  0.000000e+00  0.000298
993    21  0.0  0.000000    1  0.292748  0.000000e+00  0.000175
994    21  NaN  0.000000    1  0.320517  0.000000e+00  0.000173
995     6  0.0  0.000000    1  0.247721  0.000000e+00  0.000241
996    14  NaN  0.000000    1  0.300578  0.000000e+00  0.000175
997    11  0.0  1.000000    3  0.308052  0.000000e+00  0.000230
998    14  0.0  0.000000    1  0.313753  0.000000e+00  0.000176
999    15  0.0  0.000000    1  0.311729  0.000000e+00  0.000172
1000    4  NaN  0.600000    6  0.355934  3.515982e-06  0.000294
1001   21  NaN  0.844444   10  0.339789  3.866496e-06  0.000398
1002    1  0.0  0.000000    1  0.297983  0.000000e+00  0.000176
1003    6  0.0  0.000000    1  0.298167  0.000000e+00  0.000177
1004   22  0.0  0.000000    1  0.295975  0.000000e+00  0.000175

[1005 rows x 7 columns]

In [32]:
df_train = df[~pd.isnull(df['sal'])]
df_test = df[pd.isnull(df['sal'])]
y_train = df_train['sal']
x_train = df_train.drop('sal', axis=1)
x_test = df_test.drop('sal', axis=1)

In [33]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

clf = GradientBoostingClassifier(random_state=0)
grid_values = {'learning_rate': [0.01, 0.1],'max_depth':[4,5,6,7,8]}
grid_clf_auc = GridSearchCV(clf, param_grid = grid_values, scoring = 'roc_auc')
grid_clf_auc.fit(x_train, y_train)
print('Grid best parameter (max. AUC): ', grid_clf_auc.best_params_)
print('Grid best score (AUC): ', grid_clf_auc.best_score_)

C:\Users\wangh\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


Grid best parameter (max. AUC):  {'learning_rate': 0.1, 'max_depth': 6}
Grid best score (AUC):  0.9546710196155604


C:\Users\wangh\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:813: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


In [47]:
clf = GradientBoostingClassifier(random_state=0, learning_rate = 0.1,max_depth = 6)
clf.fit(x_train, y_train)
y_predict = clf.predict_proba(x_test)[:,1]
pd.Series(y_predict, index=df_test.index)

1       0.000436
2       0.998698
5       0.998634
8       0.066602
14      0.003590
18      0.011567
27      0.017514
30      0.986568
31      0.065572
34      0.001809
37      0.001746
40      0.001437
45      0.000493
54      0.159930
55      0.902710
60      0.009107
62      0.998634
65      0.998555
77      0.000936
79      0.001520
97      0.001089
101     0.000358
103     0.679736
108     0.001082
113     0.003124
122     0.000143
141     0.009560
142     0.998634
144     0.001041
145     0.914186
          ...   
913     0.000978
914     0.000877
915     0.000193
918     0.002502
923     0.001418
926     0.003501
931     0.000589
934     0.000173
939     0.000165
944     0.000173
945     0.001418
947     0.002308
950     0.002930
951     0.001157
953     0.000209
959     0.000157
962     0.000156
963     0.002750
968     0.001675
969     0.003008
974     0.001430
984     0.000173
987     0.001533
989     0.002313
991     0.001583
992     0.000157
994     0.000173
996     0.0001

### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [8]:
def salary_predictions():
    
    # Your Code Here
    
    return # Your Answer Here

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [70]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.info()

<class 'pandas.core.frame.DataFrame'>
Index: 488446 entries, (6, 840) to (75, 101)
Data columns (total 1 columns):
Future Connection    366334 non-null float64
dtypes: float64(1)
memory usage: 7.5+ MB


Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [97]:
df2 = future_connections.copy()
df2['jaccard_coefficient'] = list(map(lambda x: list(nx.jaccard_coefficient(G, [x]))[0][2], df2.index))
df2['resource_allocation'] = list(map(lambda x: list(nx.resource_allocation_index(G, [x]))[0][2], df2.index))
df2['pref_attachment'] = list(map(lambda x: list(nx.preferential_attachment(G, [x]))[0][2], df2.index))
df2['community_neighbors'] = list(map(lambda x: list(nx.cn_soundarajan_hopcroft(G, [x], community='Department'))[0][2], df2.index))

Future Connection  common_neighbers  jaccard_coefficient  \
(6, 840)                  0.0                 9             0.073770   
(4, 197)                  0.0                 2             0.015504   
(620, 979)                0.0                 0             0.000000   
(519, 872)                0.0                 2             0.060606   
(382, 423)                0.0                 0             0.000000   
(97, 226)                 1.0                 4             0.048193   
(349, 905)                0.0                 0             0.000000   
(429, 860)                0.0                 0             0.000000   
(309, 989)                0.0                 0             0.000000   
(468, 880)                0.0                 1             0.019608   
(228, 715)                0.0                 0             0.000000   
(397, 488)                0.0                 0             0.000000   
(253, 570)                0.0                 0             0.000000   
(435, 791)                0.0                 1             0.050000   
(711, 737)                0.0                 0             0.000000   
(263, 884)                0.0                 0             0.000000   
(342, 473)                1.0                12             0.070588   
(523, 941)                0.0                 0             0.000000   
(157, 189)                0.0                 5             0.033784   
(542, 757)                0.0                 0             0.000000   
(731, 870)                0.0                 0             0.000000   
(497, 973)                0.0                 0             0.000000   
(93, 398)                 0.0                 0             0.000000   
(102, 604)                0.0                 1             0.034483   
(206, 303)                1.0                 5             0.031847   
(57, 447)                 0.0                 1             0.010417   
(417, 758)                0.0                 0             0.000000   
(834, 837)                0.0                 1             0.500000   
(261, 557)                0.0                 4             0.063492   
(514, 740)                0.0                 1             0.071429   
...                       ...               ...                  ...   
(672, 848)                NaN                 0             0.000000   
(28, 127)                 NaN                14             0.118644   
(202, 661)                NaN                 0             0.000000   
(54, 195)                 NaN                23             0.365079   
(295, 864)                NaN                 0             0.000000   
(814, 936)                NaN                 0             0.000000   
(839, 874)                NaN                 0             0.000000   
(139, 843)                NaN                 0             0.000000   
(461, 544)                NaN                 0             0.000000   
(68, 487)                 NaN                 0             0.000000   
(622, 932)                NaN                 0             0.000000   
(504, 936)                NaN                 1             0.014706   
(479, 528)                NaN                 0             0.000000   
(186, 670)                NaN                 0             0.000000   
(90, 395)                 NaN                 5             0.078125   
(329, 521)                NaN                 2             0.018018   
(127, 218)                NaN                 6             0.067416   
(463, 993)                NaN                 0             0.000000   
(123, 142)                NaN                12             0.073620   
(764, 885)                NaN                 0             0.000000   
(144, 824)                NaN                 0             0.000000   
(742, 985)                NaN                 0             0.000000   
(506, 684)                NaN                 0             0.000000   
(505, 916)                NaN                 0     

In [100]:
df_train = df2[~pd.isnull(df2['Future Connection'])]
df_test = df2[pd.isnull(df2['Future Connection'])]
y_train = df_train['Future Connection']
x_train = df_train.drop('Future Connection', axis=1)
x_test = df_test.drop('Future Connection', axis=1)

In [104]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

clf = GradientBoostingClassifier(random_state=0)
grid_values = {'learning_rate': [0.01, 0.1],'max_depth':[4,5,6,7,8]}
grid_clf_auc = GridSearchCV(clf, param_grid = grid_values, scoring = 'roc_auc')
grid_clf_auc.fit(x_train, y_train)
print('Grid best parameter (max. AUC): ', grid_clf_auc.best_params_)
print('Grid best score (AUC): ', grid_clf_auc.best_score_)

C:\Users\wangh\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


Grid best parameter (max. AUC):  {'learning_rate': 0.1, 'max_depth': 4}
Grid best score (AUC):  0.9131046805794811


In [ ]:
def new_connections_predictions():
    
    # Your Code Here
    
    return # Your Answer Here